File: AI Trivia/Sources/ViewModels/GameViewModel.swift
import Foundation
import SwiftUI

enum GamePhase {
    case titleScreen
    case countdown
    case question
    case revealAnswer
    case gameOver
}

@Observable
class GameViewModel {
    // MARK: - Game State
    var phase: GamePhase = .titleScreen
    var currentQuestion: TriviaQuestion?
    var selectedAnswerIndex: Int?
    var score: Int = 0
    var streak: Int = 0
    var bestStreak: Int = 0
    var questionsAnswered: Int = 0
    var correctAnswers: Int = 0
    var timeRemaining: Double = 10.0
    var countdownValue: Int = 3
    var showFunFact: Bool = false
    var lastPointsEarned: Int = 0
    var comboMultiplier: Double = 1.0
    
    // Snarky host messages
    var hostMessage: String = ""
    
    var maxTimeForDisplay: Double { maxTime }
    
    // MARK: - Private
    private var usedQuestionIds: Set<Int> = []
    private var triviaBank: TriviaBank
    private var timer: Timer?
    private var countdownTimer: Timer?
    private var maxTime: Double = 10.0
    private var countdownStart: Int = 3
    private var maxCombo: Double = 4.0
    private var streakIncrement: Double = 0.5
    private var difficultyMultipliers: [String: Double] = ["easy": 1.0, "medium": 1.5, "hard": 2.0]
    private var enabledCategories: Set<String> = Set(GameSettings.allCategories)
    
    // Snarky messages for correct answers
    private let correctMessages = [
        "Well, well, well... someone's been studying!",
        "Look at the big brain on you! 🧠",
        "Ding ding ding! We have a winner!",
        "Somebody paid attention in class!",
        "You're on FIRE! 🔥 (Metaphorically. Please don't actually be on fire.)",
        "Nailed it! Your neural network is clearly well-trained.",
        "Correct! Your gradient is definitely not vanishing.",
        "RIGHT! You're basically a fine-tuned model at this point.",
        "YES! That answer was more accurate than GPT-4's hallucinations.",
        "Boom! Knowledge bomb detonated successfully. 💣",
        "Are you cheating? Because that was suspiciously fast...",
        "Your accuracy is approaching 1.0! Keep it up!",
    ]
    
    // Snarky messages for wrong answers
    private let wrongMessages = [
        "Ooof. That one's gonna leave a mark. 😬",
        "Wrong! But hey, even GPT hallucinates sometimes.",
        "Nope! Your loss function just spiked dramatically.",
        "Incorrect! Time to retrain that mental model.",
        "Swing and a miss! Your confidence was high but your accuracy... not so much.",
        "WRONG! That answer had more errors than unhandled exceptions.",
        "Yikes. Even a random forest would've gotten that one right.",
        "Nah. That's what we call 'catastrophic forgetting' in the biz.",
        "Not even close! Were you trained on corrupted data?",
        "Wrong! But don't worry, this is just your validation set. 📉",
        "Oops! Your attention mechanism clearly wasn't paying attention.",
        "Incorrect! That's going straight into your error log.",
    ]
    
    // Timeout messages
    private let timeoutMessages = [
        "Time's up! Your inference speed needs work. ⏰",
        "Too slow! Even a CPU-only model is faster than you.",
        "Tick tock! The timer waited for no one.",
        "TIME OUT! Your latency is unacceptable in production.",
        "Gone! That question has left the chat. 👋",
        "Expired! Like a deprecated API endpoint.",
    ]
    
    init() {
        self.triviaBank = TriviaBank.load()
    }
    
    // MARK: - Settings
    
    func applySettings(_ settings: GameSettings) {
        maxTime = settings.questionTime
        countdownStart = settings.countdownDuration
        maxCombo = settings.maxComboMultiplier
        streakIncrement = settings.streakBonusIncrement
        difficultyMultipliers = [
            "easy": settings.easyMultiplier,
            "medium": settings.mediumMultiplier,
            "hard": settings.hardMultiplier
        ]
        enabledCategories = settings.enabledCategories
    }
    
    // MARK: - Game Flow
    
    func startGame() {
        score = 0
        streak = 0
        bestStreak = 0
        questionsAnswered = 0
        correctAnswers = 0
        usedQuestionIds = []
        comboMultiplier = 1.0
        hostMessage = "Let's see what you've got, hotshot! 🎯"
        startCountdown()
    }
    
    private func startCountdown() {
        phase = .countdown
        countdownValue = countdownStart
        countdownTimer?.invalidate()
        countdownTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] timer in
            guard let self else { timer.invalidate(); return }
            DispatchQueue.main.async {
                self.countdownValue -= 1
                if self.countdownValue <= 0 {
                    timer.invalidate()
                    self.nextQuestion()
                }
            }
        }
    }
    
    func nextQuestion() {
        selectedAnswerIndex = nil
        showFunFact = false
        lastPointsEarned = 0
        
        if let question = triviaBank.randomQuestion(excluding: usedQuestionIds, categories: enabledCategories) {
            currentQuestion = question
            usedQuestionIds.insert(question.id)
            timeRemaining = maxTime
            phase = .question
            startTimer()
        } else {
            // All questions exhausted, reset and continue
            usedQuestionIds = []
            if let question = triviaBank.randomQuestion(excluding: usedQuestionIds, categories: enabledCategories) {
                currentQuestion = question
                usedQuestionIds.insert(question.id)
                timeRemaining = maxTime
                phase = .question
                startTimer()
            }
        }
    }
    
    func selectAnswer(_ index: Int) {
        guard phase == .question, selectedAnswerIndex == nil else { return }
        
        stopTimer()
        selectedAnswerIndex = index
        questionsAnswered += 1
        
        let isCorrect = index == currentQuestion?.correctIndex
        
        if isCorrect {
            // Score based on time remaining - faster = more points
            let timeBonus = Int(timeRemaining * 100)
            let difficultyMultiplier = difficultyMultipliers[currentQuestion?.difficulty ?? "easy"] ?? 1.0
            
            streak += 1
            if streak > bestStreak { bestStreak = streak }
            
            // Combo multiplier increases with streak
            comboMultiplier = min(maxCombo, 1.0 + Double(streak - 1) * streakIncrement)
            
            let points = Int(Double(timeBonus) * difficultyMultiplier * comboMultiplier)
            lastPointsEarned = points
            score += points
            correctAnswers += 1
            
            hostMessage = correctMessages.randomElement() ?? "Correct!"
            if streak >= 3 {
                hostMessage += " 🔥 \(streak)x STREAK!"
            }
        } else {
            streak = 0
            comboMultiplier = 1.0
            lastPointsEarned = 0
            hostMessage = wrongMessages.randomElement() ?? "Wrong!"
        }
        
        phase = .revealAnswer
        showFunFact = true
    }
    
    func timeExpired() {
        guard phase == .question else { return }
        
        stopTimer()
        questionsAnswered += 1
        streak = 0
        comboMultiplier = 1.0
        lastPointsEarned = 0
        hostMessage = timeoutMessages.randomElement() ?? "Time's up!"
        phase = .revealAnswer
        showFunFact = true
    }
    
    func endGame() {
        stopTimer()
        countdownTimer?.invalidate()
        phase = .gameOver
        
        let accuracy = questionsAnswered > 0 ? Int((Double(correctAnswers) / Double(questionsAnswered)) * 100) : 0
        if accuracy >= 80 {
            hostMessage = "Impressive! You clearly know your stuff. Final score: \(score) 🏆"
        } else if accuracy >= 50 {
            hostMessage = "Not bad! Room for improvement though. Final score: \(score)"
        } else {
            hostMessage = "Oof. Maybe review those lecture notes? Final score: \(score) 📚"
        }
    }
    
    func returnToTitle() {
        stopTimer()
        countdownTimer?.invalidate()
        phase = .titleScreen
        hostMessage = ""
    }
    
    // MARK: - Timer
    
    private func startTimer() {
        timer?.invalidate()
        timer = Timer.scheduledTimer(withTimeInterval: 0.05, repeats: true) { [weak self] timer in
            guard let self else { timer.invalidate(); return }
            DispatchQueue.main.async {
                self.timeRemaining -= 0.05
                if self.timeRemaining <= 0 {
                    self.timeRemaining = 0
                    self.timeExpired()
                }
            }
        }
    }
    
    private func stopTimer() {
        timer?.invalidate()
        timer = nil
    }
}

File: AITrivia.WinUI/ViewModels/GameViewModel.cs
using AITrivia.Models;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using Microsoft.UI.Xaml;

namespace AITrivia;

public enum GamePhase { TitleScreen, Countdown, Question, RevealAnswer, GameOver }

public partial class GameViewModel : ObservableObject
{
    // ── Game State ──────────────────────────────────────────────────────────
    [ObservableProperty] private GamePhase _phase = GamePhase.TitleScreen;
    [ObservableProperty] private TriviaQuestion? _currentQuestion;
    [ObservableProperty] private int? _selectedAnswerIndex;
    [ObservableProperty] private int _score;
    [ObservableProperty] private int _streak;
    [ObservableProperty] private int _bestStreak;
    [ObservableProperty] private int _questionsAnswered;
    [ObservableProperty] private int _correctAnswers;
    [ObservableProperty] private double _timeRemaining = 10.0;
    [ObservableProperty] private int _countdownValue = 3;
    [ObservableProperty] private bool _showFunFact;
    [ObservableProperty] private int _lastPointsEarned;
    [ObservableProperty] private double _comboMultiplier = 1.0;
    [ObservableProperty] private string _hostMessage = "";

    public double MaxTimeForDisplay => _maxTime;

    // ── Private ─────────────────────────────────────────────────────────────
    private HashSet<int> _usedQuestionIds = new();
    private readonly TriviaBank _triviaBank;
    private DispatcherTimer _questionTimer;
    private DispatcherTimer _countdownTimer;

    private double _maxTime = 10.0;
    private int _countdownStart = 3;
    private double _maxCombo = 4.0;
    private double _streakIncrement = 0.5;
    private Dictionary<string, double> _difficultyMultipliers = new() { ["easy"] = 1.0, ["medium"] = 1.5, ["hard"] = 2.0 };
    private HashSet<string> _enabledCategories = new(GameSettings.AllCategories);

    private static readonly string[] CorrectMessages =
    [
        "Well, well, well... someone's been studying!",
        "Look at the big brain on you! 🧠",
        "Ding ding ding! We have a winner!",
        "Somebody paid attention in class!",
        "You're on FIRE! 🔥 (Metaphorically. Please don't actually be on fire.)",
        "Nailed it! Your neural network is clearly well-trained.",
        "Correct! Your gradient is definitely not vanishing.",
        "RIGHT! You're basically a fine-tuned model at this point.",
        "YES! That answer was more accurate than GPT-4's hallucinations.",
        "Boom! Knowledge bomb detonated successfully. 💣",
        "Are you cheating? Because that was suspiciously fast...",
        "Your accuracy is approaching 1.0! Keep it up!",
    ];

    private static readonly string[] WrongMessages =
    [
        "Ooof. That one's gonna leave a mark. 😬",
        "Wrong! But hey, even GPT hallucinates sometimes.",
        "Nope! Your loss function just spiked dramatically.",
        "Incorrect! Time to retrain that mental model.",
        "Swing and a miss! Your confidence was high but your accuracy... not so much.",
        "WRONG! That answer had more errors than unhandled exceptions.",
        "Yikes. Even a random forest would've gotten that one right.",
        "Nah. That's what we call 'catastrophic forgetting' in the biz.",
        "Not even close! Were you trained on corrupted data?",
        "Wrong! But don't worry, this is just your validation set. 📉",
        "Oops! Your attention mechanism clearly wasn't paying attention.",
        "Incorrect! That's going straight into your error log.",
    ];

    private static readonly string[] TimeoutMessages =
    [
        "Time's up! Your inference speed needs work. ⏰",
        "Too slow! Even a CPU-only model is faster than you.",
        "Tick tock! The timer waited for no one.",
        "TIME OUT! Your latency is unacceptable in production.",
        "Gone! That question has left the chat. 👋",
        "Expired! Like a deprecated API endpoint.",
    ];

    public GameViewModel()
    {
        AppLog.Info("GameViewModel ctor: begin");
        _triviaBank = TriviaBank.Load();

        _questionTimer = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(50) };
        _questionTimer.Tick += OnQuestionTimerTick;

        _countdownTimer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(1) };
        _countdownTimer.Tick += OnCountdownTimerTick;
        AppLog.Info("GameViewModel ctor: timers created");
    }

    // ── Settings ─────────────────────────────────────────────────────────────
    public void ApplySettings(GameSettings settings)
    {
        AppLog.Info($"GameViewModel.ApplySettings questionTime={settings.QuestionTime} countdown={settings.CountdownDuration} categories={settings.EnabledCategories.Count}");
        _maxTime = settings.QuestionTime;
        _countdownStart = settings.CountdownDuration;
        _maxCombo = settings.MaxComboMultiplier;
        _streakIncrement = settings.StreakBonusIncrement;
        _difficultyMultipliers = new Dictionary<string, double>
        {
            ["easy"] = settings.EasyMultiplier,
            ["medium"] = settings.MediumMultiplier,
            ["hard"] = settings.HardMultiplier
        };
        _enabledCategories = new HashSet<string>(settings.EnabledCategories);
    }

    // ── Game Flow ─────────────────────────────────────────────────────────────
    [RelayCommand]
    public void StartGame()
    {
        AppLog.Info("GameViewModel.StartGame begin");
        Score = 0; Streak = 0; BestStreak = 0;
        QuestionsAnswered = 0; CorrectAnswers = 0;
        _usedQuestionIds = new();
        ComboMultiplier = 1.0;
        HostMessage = "Let's see what you've got, hotshot! 🎯";
        StartCountdown();
        AppLog.Info("GameViewModel.StartGame end");
    }

    private void StartCountdown()
    {
        AppLog.Info($"GameViewModel.StartCountdown startValue={_countdownStart}");
        Phase = GamePhase.Countdown;
        CountdownValue = _countdownStart;
        _countdownTimer.Stop();
        _countdownTimer.Start();
    }

    private void OnCountdownTimerTick(object? sender, object e)
    {
        AppLog.Info($"GameViewModel.OnCountdownTimerTick before={CountdownValue}");
        CountdownValue--;
        if (CountdownValue <= 0)
        {
            _countdownTimer.Stop();
            AppLog.Info("GameViewModel.OnCountdownTimerTick transitioning to NextQuestion");
            NextQuestion();
        }
    }

    [RelayCommand]
    public void NextQuestion()
    {
        AppLog.Info("GameViewModel.NextQuestion begin");
        SelectedAnswerIndex = null;
        ShowFunFact = false;
        LastPointsEarned = 0;

        var question = _triviaBank.RandomQuestion(_usedQuestionIds, _enabledCategories);
        if (question is null) { _usedQuestionIds.Clear(); question = _triviaBank.RandomQuestion(_usedQuestionIds, _enabledCategories); }
        if (question is null)
        {
            AppLog.Info("GameViewModel.NextQuestion no question available");
            return;
        }

        CurrentQuestion = question;
        _usedQuestionIds.Add(question.Id);
        TimeRemaining = _maxTime;
        Phase = GamePhase.Question;
        StartTimer();
        AppLog.Info($"GameViewModel.NextQuestion end id={question.Id}");
    }

    [RelayCommand]
    public void SelectAnswer(int index)
    {
        if (Phase != GamePhase.Question || SelectedAnswerIndex.HasValue) return;

        StopTimer();
        SelectedAnswerIndex = index;
        QuestionsAnswered++;

        var isCorrect = index == CurrentQuestion?.CorrectIndex;
        if (isCorrect)
        {
            var timeBonus = (int)(TimeRemaining * 100);
            var diffMult = _difficultyMultipliers.GetValueOrDefault(CurrentQuestion?.Difficulty ?? "easy", 1.0);
            Streak++;
            if (Streak > BestStreak) BestStreak = Streak;
            ComboMultiplier = Math.Min(_maxCombo, 1.0 + (Streak - 1) * _streakIncrement);
            var points = (int)(timeBonus * diffMult * ComboMultiplier);
            LastPointsEarned = points;
            Score += points;
            CorrectAnswers++;
            HostMessage = CorrectMessages[Random.Shared.Next(CorrectMessages.Length)];
            if (Streak >= 3) HostMessage += $" 🔥 {Streak}x STREAK!";
        }
        else
        {
            Streak = 0;
            ComboMultiplier = 1.0;
            LastPointsEarned = 0;
            HostMessage = WrongMessages[Random.Shared.Next(WrongMessages.Length)];
        }

        Phase = GamePhase.RevealAnswer;
        ShowFunFact = true;
    }

    private void TimeExpired()
    {
        if (Phase != GamePhase.Question) return;
        StopTimer();
        QuestionsAnswered++;
        Streak = 0;
        ComboMultiplier = 1.0;
        LastPointsEarned = 0;
        HostMessage = TimeoutMessages[Random.Shared.Next(TimeoutMessages.Length)];
        Phase = GamePhase.RevealAnswer;
        ShowFunFact = true;
    }

    [RelayCommand]
    public void EndGame()
    {
        StopTimer();
        _countdownTimer.Stop();
        Phase = GamePhase.GameOver;
        var accuracy = QuestionsAnswered > 0
            ? (int)(CorrectAnswers * 100.0 / QuestionsAnswered) : 0;
        HostMessage = accuracy >= 80
            ? $"Impressive! You clearly know your stuff. Final score: {Score} 🏆"
            : accuracy >= 50
                ? $"Not bad! Room for improvement though. Final score: {Score}"
                : $"Oof. Maybe review those lecture notes? Final score: {Score} 📚";
    }

    [RelayCommand]
    public void ReturnToTitle()
    {
        StopTimer();
        _countdownTimer.Stop();
        Phase = GamePhase.TitleScreen;
        HostMessage = "";
    }

    // ── Timer ─────────────────────────────────────────────────────────────────
    private void StartTimer()
    {
        _questionTimer.Stop();
        _questionTimer.Start();
        AppLog.Info("GameViewModel.StartTimer");
    }

    private void StopTimer() => _questionTimer.Stop();

    private void OnQuestionTimerTick(object? sender, object e)
    {
        TimeRemaining -= 0.05;
        if (TimeRemaining <= 0)
        {
            TimeRemaining = 0;
            TimeExpired();
        }
    }
}
